### PR TITLE
feat: add support for custom start/end times

### PR DESCRIPTION
**Pull request by watson**
Both transaction and spans can now have their default start and end times overridden by providing a custom timestamp.

The custom timestamp is a Unix Time Stamp representing the number of milliseconds since January 1, 1970, 00:00:00 UTC. Sub-millisecond precision can be achieved using decimals.

Ability to set custom start time have been added to:

- `agent.startTransaction()`
- `agent.startSpan()`
- `transaction.startSpan()`

Ability to set custom end time have been added to:

- `agent.endTransaction()`
- `transaction.end()`
- `span.end()`

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update documentation

### CODE REVIEW

File: docs/transaction-api.asciidoc [hunk 1]
```diff
@@ -69,6 +69,14 @@ Defaults to `unnamed`
 You can alternatively set this via <<span-type,`span.type`>>.
 Defaults to `custom`
 
+* `options` - The following options are supported:
+
+** `startTime` - The time when the span started.
```
Comment (watson) [inline docs/transaction-api.asciidoc]: I didn't document the ability to set `traceparent` here as we hadn't done that previously. LMKWYT

File: test/agent.js [hunk 2]
```diff
@@ -15,12 +15,50 @@ var config = require('../lib/config')
...
+  })
+
+  t.test('options.traceparent', function (t) {
+    var agent = Agent()
+    agent.start()
+    var traceparent = '00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01'
+    var trans = agent.startTransaction('foo', 'bar', { traceparent })
+    t.equal(trans._context.version, '00')
+    t.equal(trans._context.traceId, '4bf92f3577b34da6a3ce929d0e0e4736')
+    t.notEqual(trans._context.id, '00f067aa0ba902b7')
```
Comment (Qard) [inline test/agent.js]: Maybe also verify the `parentId`?

File: test/agent.js [hunk 3]
```diff
@@ -15,12 +15,50 @@ var config = require('../lib/config')
...
+  })
+
+  t.test('traceparent (legacy)', function (t) {
+    var agent = Agent()
+    agent.start()
+    var traceparent = '00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01'
+    var trans = agent.startTransaction('foo', 'bar', traceparent)
+    t.equal(trans._context.version, '00')
+    t.equal(trans._context.traceId, '4bf92f3577b34da6a3ce929d0e0e4736')
+    t.notEqual(trans._context.id, '00f067aa0ba902b7')
```
Comment (Qard) [inline test/agent.js]: Maybe also verify the `parentId`?

File: test/agent.js [hunk 4]
```diff
@@ -143,6 +192,45 @@ test('#startSpan()', function (t) {
...
+
+  t.test('options.traceparent', function (t) {
+    var agent = Agent()
+    agent.start()
+    agent.startTransaction()
+    var traceparent = '00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01'
+    var span = agent.startSpan(null, null, { traceparent })
+    t.equal(span._context.version, '00')
+    t.equal(span._context.traceId, '4bf92f3577b34da6a3ce929d0e0e4736')
+    t.notEqual(span._context.id, '00f067aa0ba902b7')
```
Comment (Qard) [inline test/agent.js]: Maybe also verify the `parentId`?

File: test/agent.js [hunk 5]
```diff
@@ -143,6 +192,45 @@ test('#startSpan()', function (t) {
...
+
+  t.test('traceparent (legacy)', function (t) {
+    var agent = Agent()
+    agent.start()
+    agent.startTransaction()
+    var traceparent = '00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01'
+    var span = agent.startSpan(null, null, traceparent)
+    t.equal(span._context.version, '00')
+    t.equal(span._context.traceId, '4bf92f3577b34da6a3ce929d0e0e4736')
+    t.notEqual(span._context.id, '00f067aa0ba902b7')
```
Comment (Qard) [inline test/agent.js]: Maybe also verify the `parentId`?

File: test/agent.js (hunk 2 above)
Comment (watson) [reply test/agent.js]: Fixed

File: test/agent.js (hunk 3 above)
Comment (watson) [reply test/agent.js]: Fixed

File: test/agent.js (hunk 4 above)
Comment (watson) [reply test/agent.js]: Fixed

File: test/agent.js (hunk 5 above)
Comment (watson) [reply test/agent.js]: Fixed

File: lib/instrumentation/transaction.js [hunk 6]
```diff
@@ -93,7 +93,26 @@ Transaction.prototype.startSpan = function (name, type) {
...
+    opts = {
+      timer: this._timer,
+      traceparent: opts
+    }
+  } else {
+    opts.timer = this._timer
+    if (!opts.traceparent) {
+      opts.traceContext = (this._agent._instrumentation.activeSpan || this)._context
+    }
+  }
```
Comment (Qard) [inline lib/instrumentation/transaction.js]: Maybe move this logic _into_ the `Span` constructor? It still has access to the transaction there, and that would mean the tests don't need to explicitly pass the `traceContext` in everywhere in the span test file.

Comment (watson) [reply lib/instrumentation/transaction.js]: Fixed

**Pull request merged**